### ISSUE TITLE
Can't aggregate Salesforce Files (ContentDocumentLink) as child object

### ISSUE THREAD

**Issue by matthewsouther** [afawcett/declarative-lookup-rollup-summaries#433]
Due to an apparent SOQL limitation with the ContentDocumentLink standard object, attempts to use DLRS to capture the ID of the most recent file on a parent record fail.

The validation error reads:

_"Apex trigger dlrs_ContentDocumentLinkTrigger caused an unexpected exception, contact your administrator: dlrs_ContentDocumentLinkTrigger: execution of AfterInsert caused by: System.QueryException: Implementation restriction: ContentDocumentLink requires a filter by a single Id on ContentDocumentId or LinkedEntityId using the equals operator or multiple Id's using the IN operator.: Class.dlrs.LREngine.QueryExecutor.query: line 631, column 1"_

See Success Community thread: https://success.salesforce.com/0D53A00002v7xXO
![11-16-2016 10-50-18 am](https://cloud.githubusercontent.com/assets/22968569/20437595/6eaa9af2-ad69-11e6-91ca-a0015ee2d913.png)

**Comment by afawcett** [comment]
Thanks for sharing @matthewsouther, i'll flag this as a future enhancement, maybe there is a workaround the platform limitation.

**Comment by tmiller910** [comment]
Are there any updates to this thread? I'm trying to count the number of files for a custom object and run into this if I try to run via Process Builder or Real-Time. When I just run the lookup to calculate from the Manage screen, it runs through and updates.

**Comment by DouglasCAyers** [comment]
I think this is fixed per pull request https://github.com/afawcett/declarative-lookup-rollup-summaries/pull/585

**Comment by afawcett** [comment]
Yes @DouglasCAyers i would agree.

**Comment by afawcett** [comment]
Fixed in v2.10

**Comment by rishikumar2018** [comment]
Hi,
We tried to implement this for attachments it is working fine when we are adding the attachment but when we delete the attachment it is not working as expected. We are trying to put the counts of attachments related to contract object. Count gets increased when we are adding the attachment but it is not decreased when we deleted the attachment.

Can you please help over it.

**Comment by SebasCanseco** [comment]
Hi @afawcett, I have tested this with v2.10 and I'm not getting reliable results:
1) When I upload a file to a Contact once, the DLRS doesn't work, but when I immediately upload a second file it gets calculated but doesn't count the one of the files, so we are off the correct value by one.
2) As @rishikumar2018 stated, I'm not able to get this to work when deleting a file from Contact. The Aggregate Result Field in the parent object (Contact) doesn't get updated.

Any clues? Thanks!

**Comment by afawcett** [comment]
Ok thanks for confirming, i'll re-open since our hope this was fixed via https://github.com/afawcett/declarative-lookup-rollup-summaries/pull/585 does not seem to have transpired sadly.

**Comment by Ltowns** [comment]
I'm unable to deploy the trigger for the child object ContentDocument, receive the following error: Error:
dlrs_ContentDocumentTest.testTrigger System.TypeException: DML operation INSERT not allowed on ContentDocument Class.dlrs.RollupService.testHandler: line 282, column 1 Class.dlrs_ContentDocumentTest.testTrigger: line 11, column 1. This is for a rollup counting ContentDocument on Account. Is this related to these reported issues with ContentDocumentLink?

**Comment by DouglasCAyers** [comment]
Hi @Ltowns,

You should use ContentDocumentLink object instead of ContentDocument or ContentVersion objects. 

One reason is that sharing security is based on ContentDocumentLink object. Users, even admins, only can query and see ContentDocument or ContentVersion records explicitly shared to them. Any whose access is inferred by the user having access to the record the file has been shared to (e.g. account or contact) are ignored. So you would not get accurate counts and the counts would change based on the context user each time. 

The other reason is that ContentDocument records can’t be inserted or updated directly which is why the auto-generated test fails. The document is created by Salesforce when a ContentVersion record is inserted and its ContentDocumentId field is null.

**Comment by Ltowns** [comment]
Got it, thanks. Then it appears that by using ContentDocumentLink I can't
filter by the Title or FileType. What I'm trying to do is identify if there
is a File attached to an account that contains the title "Account Form" by
getting a count using that as a filter.



Laurie Townsend
Sr. Consultant
Groundswell Cloud Solutions, Inc.
groundswell.force.com | 707-321-9931 | laurie@gscloudsolutions.com |







CONFIDENTIALITY NOTICE: The contents of this email message and any
attachments are confidential and intended solely for the addressee(s) and
may be legally protected from disclosure. If you are not the intended
recipient of this message or their agent, or if this message has been
addressed to you in error, please immediately alert the sender by reply
email and then delete this message and any attachments. If you are not the
intended recipient, you are hereby notified that any use, dissemination,
copying, or storage of this message or its attachments is strictly
prohibited.

**Comment by DouglasCAyers** [comment]
You might try cross object filter criteria like “ContentDocument.Title = ‘Account Form’”

**Comment by Jesus-GA** [comment]
Hello @afawcett,

Please, could you let us know if there is an estimated time of resolution for the issue reported by Sebas on the 31st of January?

Many thanks!
Jesus

**Comment by mkolodner** [comment]
I'm also finding that the count in Realtime does not go down for linked Files (rollup based on ContentDocumentLink), though it goes up just fine. A Scheduled rollup does result in number going down for deleted files. But for some use cases a scheduled rollup may not meet the need.

**Comment by SebasCanseco** [comment]
Hi, I was wondering if this has been fixed. I'm working with another client and they could really use DLRS as long as it supports ContentDocumentLink object. Thanks for the hard work!

**Comment by jwallace2040** [comment]
Hi, I'm trying to count the number of files attached to a record. I'm getting this error, and I'm stuck tracking down the issue.  Any help or guidance is appreciated.

![dlrs_fistula2](https://user-images.githubusercontent.com/77695597/105391672-5a45db80-5be8-11eb-9d62-ec4aad06d105.JPG)
![dlrs_fistula1](https://user-images.githubusercontent.com/77695597/105391674-5ade7200-5be8-11eb-8c44-cab0dd1fe64c.JPG)

**Comment by CathyJohnson-silverline** [comment]
This issue where the count of files will correctly count for added files but not for deleted files is still occurring in the latest version (2.13). I noticed that if I select the option for "Remove from Record" then the decrementing on the count does work, but not for a Delete. We need this to be accurate in real time so, unfortunately, a scheduled run to correct any incorrect counts won't work. 

I would love to understand why the behavior on this object is different from other Salesforce objects. 

![RelatedListForFilesOptions](https://user-images.githubusercontent.com/84247424/118375646-9aaada00-b588-11eb-9ca4-be0acc538453.png)

**Comment by aheber** [comment]
@JobbyC, are you able to collect any logging information for what is happening? Maybe use Salesforce Debug Logs and send us what you find there? Unfortunately that error message doesn't include any details.

Can you also include the details of your rollup so we can understand how you have it configured?

Completely random guess but I'd suggest one of two likely scenarios. Something in the rollup is configured incorrectly which is throwing an error or something about your opportunities is rejecting this record edit to save the rollup value, maybe a validation rule or something similar.

For future reference I'll do what I can here but I recommend you check into the Trailhead community group here, https://trailhead.salesforce.com/trailblazer-community/groups/0F9300000009O5pCAE. You're much more likely to get help and support there.

Thanks!

**Comment by SF-Q** [comment]
Hi, Has the issue been fixed to reduce file count when a file is deleted? Are there any workarounds to address this?

**Comment by JobbyC** [comment]
@SF-Q Yes managed to resolve it. I attached the roll-up that I did here, hope this help :)
![image](https://user-images.githubusercontent.com/51979592/147564235-dec15294-41cc-4b1c-a177-730706e085a3.png)

**Comment by SF-Q** [comment]
> @SF-Q Yes managed to resolve it. I attached the roll-up that I did here, hope this help :) ![image](https://user-images.githubusercontent.com/51979592/147564235-dec15294-41cc-4b1c-a177-730706e085a3.png)

I have the same configuration, the count increases when I upload files however the count does decrease when I Delete the file.

**Comment by SF-Q** [comment]
> > @SF-Q Yes managed to resolve it. I attached the roll-up that I did here, hope this help :) ![image](https://user-images.githubusercontent.com/51979592/147564235-dec15294-41cc-4b1c-a177-730706e085a3.png)
> 
> I have the same configuration, the count increases when I upload files however the count does decrease when I Delete the file.

Does the file count decrease when you delete the files? Did you have to write additional code?

**Comment by SF-Q** [comment]
@JobbyC - I have the same configuration and the file count does not decrease when I delete files. Did you have to write additional code?

**Comment by JobbyC** [comment]
@SF-Q If you use the delete option, it will not decrease the number. So, need to click the last option "Removed from Record" and the recalculation will work. 
![image](https://user-images.githubusercontent.com/51979592/147574561-50ef7035-83c8-4e97-a276-214e72a1cffd.png)

**Comment by riekusr** [comment]
I am trying to do this: 
```
SELECT Id, ContentDocumentId,  ContentDocument.LatestPublishedVersion.Vervaldatum__c , ContentDocument.LatestPublishedVersion.Soort_Document__c   FROM ContentDocumentLink
WHERE LinkedEntityId  = '0010600002ET8ZbAAL'  AND ContentDocument.LatestPublishedVersion.Soort_Document__c = 'KvK document'
```
<img width="1400" alt="afbeelding" src="https://github.com/SFDO-Community/declarative-lookup-rollup-summaries/assets/34796052/b472e250-a9be-4822-8d21-d27daf491b46">

**Comment by aheber** [comment]
@riekusr I don't think this issue ticket would be the correct place to discuss your issue. Please reach out on the Trailblazer Community Group (https://trailhead.salesforce.com/trailblazer-community/groups/0F9300000009O5pCAE) and we can discuss it there.

Thanks!
